Design.md

```diff
@@ -227,8 +229,26 @@ This API will enable or disable user on Vault.
 
 
 ### Create S3 Credential
-This API creates S3 credential for the user.
-1. `generate-access-key` api will be called using assumed role credentials.
+This API creates S3 credentials for the user.
+1. The Generate Access Key API will be called using assumed role credentials.
+1. Encrypt the secret key (For more, see [SecretKey-Encryption-Strategy](#SecretKey-Encryption-Strategy)). Store the encrypted secret key on Redis Sentinel in the hash named "osis:s3credentials", with the hash key formatted as `<Username>__<AccessKeyID>`.
+1. Log the access key ID _(Log location and tools for logging are yet to be decided)_
+
+**Important Notes:**
+* Redis Sentinel is subject to an ongoing rolling deployment.
+* If Redis Sentinel fails or crashes, the vCloud Director's OSE service must be restarted with the following command:
+    ```shell
+    $ ose service restart
+    ```
+  * Once OSE service restart process is finished,
+     * OSE will invoke `listS3Credentials` API to identify the keys for object storage operations. OSIS will return only an empty list for `listS3Credentials` API (as no key  is available for the user on Redis).
+     * If no keys are present for a user, OSE will invoke `createS3Credentials` API to create a new key for object storage operations.
+  * All access keys created by OSIS before OSE restart time (`ose-restart-time`) are dead keys on the Vault DB, which OSE/OSIS cannot use and must be manually deleted from the Vault. 
+  * A log location for all the dead access key IDs created by OSIS and a script to run manually to delete these dead keys from the Vault DB, is provided.
```

Comment:
Access keys are not necessarily dead keys, for instance, if a vCloud Director Tenant User creates access keys and puts them in his application. After Redis sentinel data lost + OSIS reboot, those keys are still used.

I suggest we give vCloud Director Tenant Users the responsibility to clean unused access keys using S3 console or IAM API because vCloud Director Tenant Users are the only ones who knows if the keys are used or not.

This design issue should be mention in the customer-facing documentation has a technical limitation due to VMware OSE/OSIS framework. 
```suggestion
  * Access keys created by OSIS before OSE restart time (`ose-restart-time`) will not be listed by OSIS. vCloud Director Tenant Users are responsible to clean up unused access keys using S3 console or IAM API because they are the only one who knows if the access keys are in used or not.
  * Access key IDs generated by OSIS will be logged
```

Reply:
I agree :)
I was thinking as well, about the "dead" word.

Reply:
@saitharunthodupunuru  I'm not sure I understand this one fully yet. 

Are we saying that any OSE restart:

1. Orphans access keys prior to the restart time
2. Orphaned keys won't be shown in vCloud Director 
3. The vCloud Director Tenant User will need to use the S3 Console or IAM API to:
    a.) Delete the unlisted (orphaned) key
    b.) Recreate a new access key (either in s3console, IAM API or vCloud Director)
4. Replace in their application the orphaned key pair (access+secret key) with the new key pair from step 3b.?

Reply:
@scality-trevorbenson 
The above design is to mention that if entire Redis cluster fails, it will lose all the data on it including the access keys created by OSIS. If that happens all those keys cannot be listed by OSIS on OSE UI. 
* OSE has some kind of cache to itself which handles the keys on it when accessed but if OSIS is not displaying the keys, there will be an mismatch in terms of keys by OSE and OSIS. To avoid such scenario, it is suggested to restart OSE if entire Redis sentinel cluster crashes to clear the cache on OSE. Once this is done, OSE automatically creates new keys as OSIS cannot return the keys which are not present on Redis.
* The keys deleted from Redis cannot be managed through OSE as OSIS cannot return them.
* The keys which are not shown on OSE by OSIS were earlier termed as `dead` can still be used in the applications which has procured them before the crash. It is mentioned here that the responsibility of the maintenance of these keys are by the tenant user.

Reply:
@terrydervaux  fixed it in e87f4cb37a396aa2237d83bf61f623d6fb24ccc4

Reply:
@saitharunthodupunuru This sounds like a redis crash of any kind could puts an extremely large amount of work in the hands of the tenant to fix, for example if tenants have tens or hundreds of keys for different apps. Is there any viable way to make a full redis restart refresh the cache of OSIS keys, therefore making it completely transparent to the tenant user the redis crash occurred?

If we can provide this functionality I think it would drastically reduce the problems presented by a redis crash. However I will understand if there is no simple way to make this transparent, and if so the rest of the PR LGTM for approval.

Reply:
The problem here is Vault does not return `secretKeys` any time after it is created so there is no way we can refresh the existing keys by OSIS on Redis if that data is lost. OSE always expects OSIS to return `secretKeys` in plaintext when needed. 

And this is the primary disadvantage when we use Redis as a Database.  As Redis is just an in-memory data store, all we need to do is to maintain Redis Master and replicas to not lose data.

I understand the ordeal is not straight forward or not simple but this is there just for a rainy day which we don't expect often